### PR TITLE
Enable linger for workshop user

### PR DESCRIPTION
**Pull request by lachypjones**
# Description
## Background
`/run/user/<uid>` gets created by pam_systemd when a user has an _active_ login shell. 
This creates a problem when an interface (ie. Desktop) looks to use this directory to proxy sockets into. 

### Linger
This PR enables lingering (`loginctl enable-linger`) for the workshop user as part of the container start script. Linger starts the user manager for `workshop` on boot, enabling interfaces that rely on resources in `$XDG_RUNTIME_DIR` to work.  

Linger introduces no modification to the system environment beyond what a normal login session does, namely:

- `/run/user/$UID` is either created or mounted
- `$XDG_SESSION_ID` is initialised
-  A new systemd scope unit is created for the session (user@1000.service, user-1000.slice)

https://www.freedesktop.org/software/systemd/man/252/pam_systemd.html

### Start Command
This PR also moves the start command to an external shell script (`start_command.sh`) to enable automated linting with CI runners. The script is embedded into workshopd at compile time.

# Self-review quick check

 * [x] Make decisions that cost a lot to reverse explicit in the PR description.
 * [x] Avoid nested conditions.
 * [x] Delete dead code and redundant comments.
 * [x] Normalise symmetries by sticking to doing identical things identically. 
 ```
 // one way to handle errors
 if err := f(); err != nil {
    ...
 }
 
 // one way to handle multiple returns
 val, err := f()
 if err != nil {
    ...
 }
 ...
 ```
 * [x] Check that coupled code elements, files, and directories are adjacent. For example, test data is stored as close as possible to a test.
 * [x] Put variable declaration and initialisation together.
 * [x] Divide large expressions into digestable and self-explanatory ones. Use multiple variables if required.
 * [x] Put a blank line between two logically different chunks of code.

### CODE REVIEW

File: internal/workshop/lxd/lxd_backend.go [hunk 1]
```diff
@@ -183,9 +187,7 @@ func (s *Backend) StartWorkshop(ctx context.Context, name string) error {
 			UserId:  0,
 			GroupId: 0,
 			Command: []string{
-				"bash", "-eu", "-c", "while " +
-					"[ \"$(systemctl is-system-running 2>/dev/null)\" != \"running\" ] && " +
-					"[ \"$(systemctl is-system-running 2>/dev/null)\" != \"degraded\" ]; do :; done",
+				"bash", "-euc", string(startCommand),
```
Comment (dmitry-lyfar) [inline internal/workshop/lxd/lxd_backend.go]: @lachypjones startCommand is already a string, why typecast here?

**Comment by dmitry-lyfar** [comment]
@jonathan-conder can you have a look into this change as well, it is important? We need side-effects of having a login session to enable something like a Desktop interface.

**Comment by jonathan-conder** [comment]
Sure, I think something like this might be useful for the camera interface too but I'll try to see if there are any reasons not to do something like this

File: internal/workshop/lxd/start_command.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+while [ "$(systemctl is-system-running 2>/dev/null)" != running ] \
```
Comment (dmitry-lyfar) [inline internal/workshop/lxd/start_command.sh:4]: @lachypjones please move the comments about the way we query systemd here (+link to the article).

**Review by dmitry-lyfar — APPROVED** [review]
looks good to me. Please note, we don't force push if comments were received for a PR: https://canonical-workshop.readthedocs-hosted.com/en/latest/contributing/#coding

**Review by jonathan-conder — APPROVED** [review]
Seems fine as far as I can tell. I think we should test the properties we want directly, e.g. does /run/user/1000 exist. But that can be covered in the interface-desktop tests.

Only relevant thing I found from searching is this TODO item which I don't really understand, but also don't think it will break anything for us:  https://github.com/systemd/systemd/commit/648b122045182a1ac19d1622817e2a350beb0354

**Comment by dmitry-lyfar** [comment]
@lachypjones this is a good idea, can we extend the existing e2e test to check that the runtime dir is set? Are you expecting `$XDG_RUNTIME_DIR` to be set for the user login session (and /run dir exists as a result) compared to when not enabled, the directory won't exist:

```
workshop exec test -- test -d /run/user/1000
```

**Comment by jonathan-conder** [comment]
One thing I noticed is linger doesn't actually persist the user session, only the slice/services, e.g.:

```console
$ workshop exec ws -- loginctl enable-linger workshop
$ workshop exec ws -- loginctl list-sessions
No sessions.
$ workshop exec ws -- ls /run/user
1000
```

**Comment by dmitry-lyfar** [comment]
Ah, yes, true, it starts the [user manager](https://www.freedesktop.org/software/systemd/man/latest/loginctl.html#:~:text=Enable%2Fdisable%20user%20lingering%20for,or%20numeric%20UIDs%20as%20argument.) which results in the dir creation: 

> Enable/disable user lingering for one or more users. If enabled for a specific user, 
> a user manager is spawned for the user at boot and kept around after logouts.

**Comment by lachypjones** [comment]
Thanks all,
I'll modify the test to check for the _resultant_ behavior of linger rather than checking whether linger is enabled (good call @jonathan-conder).

As for the user-manager commentary, do we feel my the comments on the PR don't reflect this enough?

**Comment by jonathan-conder** [comment]
> As for the user-manager commentary, do we feel my the comments on the PR don't reflect this enough?

Na I think it's fine

**Pull request merged**